### ISSUE TITLE
Drop Destination Scripts

### ISSUE THREAD

**Issue by da3dsoul** [ShokoAnime/ShokoServer#504]
I know the title is vague. Basically, we have a beautiful script engine for renaming files, and no real control over moving them.
So:
```
# DEST acts as a return statement with the final decision on where the file goes
# if statements go until DEST is reached. No real var support. I don't think we need it. if an if is evaluated as false, it skips to the line after the next DEST
# We are cool kids. Use '/' and it will convert to whatever stupid windows format
if {Restricted} == True
DEST {Drop Destination 1 (needs a way to choose them)}/{GroupName}/{SeriesName} # Series Name is the default one that goes by current setting
if {AnimeType} == "Movie" and {HasTvDBLink}
DEST {Drop Destination 2}/{GroupName}/{SeriesName_TvDB} # Specifically the TvDB one
if {AnimeType} == "Movie" and {SeriesName_Language_Type} == "Exact" # when using {SeriesName_Language_Type} in if statements, return the match accuracy as "Exact", "Language", or "None"
# Split on _ and attempt to find a title that matches, if used in DEST without an if check, use order of progession: Exact Match -> Same Language Main or Official Title -> Romaji Main Title
DEST {Drop Destination 2}/{GroupName}/{SeriesName_Language_Type}
if {Drop Destination 3} == "OK" # using a drop destination in an if statement returns the status of the folder. Useful for network drives. Possibilities are "OK", "FULL", "OFFLINE", "UNAUTHORIZED". if you ask: if {Drop 3} is "FULL", it should actually check if the file we are moving will fit
DEST {Drop 3}/Anime Overflow # Your standard network out of space drive. You can use normal text to specify a static string
```
The system we use now is a painfully simple script still available in this system:
```
DEST {The only Drop Destination, as we only support one}/{SeriesName_xjat_main}
```
@maxpiva @ElementalCrisis I have such great ideas when I sleep haha

**Comment by da3dsoul** [comment]
This will replace #50 #51 #62 #378, #98 should be included, and folders are no longer treated as source and destination, the script can move from a 'destination' to a 'source' if the criteria change.

**Comment by da3dsoul** [comment]
For example:
```
if {Dest 1} == OK
DEST {Dest 1}/{GroupName}/{SeriesName}
if {Dest 2} == OK # internal drive, always online, but need to check for space
DEST {Dest 2}/Temp/{SeriesName}
# if neither condition is met, we simply don't move it anywhere
```
If this script is run while the network drive {Dest 1} is offline, it is attempted to be put on a larger internal drive (Likely the drive you keep Tera, WoW, and FFXIII, but not enough space to keep all of Bleach, Ranma 1/2, Naruto, and One Piece). The second time this is run, though, it will gladly take the files out of {Dest 2} and move them to {Dest 1} if it is online now.

**Comment by ElementalCrisis** [comment]
I like the idea, should make the Renamer Utility a lot more useful. 

However we also need to look at a basic or standard version of the Renamer utility for new users as it's pretty confusing at first glance. I was thinking we can use buttons that the user fills out the values for.

**Comment by da3dsoul** [comment]
It's a good chance to improve both the GUIs. I can't do mockups like you...so photoshop. be back in a few with some ideas

**Comment by bigretromike** [comment]
If its possible I would add moving image files inside series dir also with those video files (covers, etc) those tend to get lost and it would be good improvement for future features

**Comment by da3dsoul** [comment]
Thanks brm. That's necessary. Detect if there is a folder with only images (and system files), if so move it completely. If there are just loose images, move them.

**Comment by bigretromike** [comment]
would love that. but as system file you refer to thumb.dll ? we can safely remove that and move rest.
Im also not sure if subtitles are moved with files.
it is rare to have them as external file but few times I had that 👍

**Comment by da3dsoul** [comment]
Yes, subs are already moved with it. By system file, I mean it will still count it as only having images, even if an ini, thumbs.db, etc exist.

**Comment by da3dsoul** [comment]
![img](https://dl.dropboxusercontent.com/u/31454410/Github%20Issues/File%20Renamer%20Mockup.png)

**Comment by ElementalCrisis** [comment]
I like it, it will improve over time. :)

**Comment by da3dsoul** [comment]
Do we have a wiki for the file renamer?

**Comment by ElementalCrisis** [comment]
Just the website.

**Comment by da3dsoul** [comment]
I updated the above post to have the current list of variables. Comment any that need added

**Comment by bigretromike** [comment]
could this support more destination folders ? ultimate drop folders update ?:-)

**Comment by da3dsoul** [comment]
Yes as many as desired

**Comment by maxpiva** [comment]
Well there is a limitation currently cloud filesystem do no support
uploading, since it's a complex thing to do :p resume capability and size
problems are lacking on some cloud providers. So have it right and
resilient is a problem right now. So you cannot move file crosscloud or
cross file-cloud

**Comment by da3dsoul** [comment]
​Ah, that sucks. Still is hella useful on local networks, though​.

**Comment by da3dsoul** [comment]
It'd be nice if we could find a nice library for it

**Comment by maxpiva** [comment]
I have to code the cloud filesystem because of the lack of such library :P 
But i'll enable uploading on all filesystems eventually. I need to do some Unit Testing, and a lot of error handling for each cloud which is a pain in the a?s. Amazon Cloud Has a lots of problems and is probably the most used. It has size timeout problems, and v1 dosn't support resume, you imagine uploading a 15Gb -30Gb file without resume capabilities?. v2 seems to support it, but Amazon rejected my petition for beta. I could eventually reverse engineering they own client. But again is time. You can guess why Plex Cloud with Amazon not surfaces yet ;) Amazon Cloud Api is not stable enough.

**Comment by AreFinstad** [comment]
@da3dsoul its missing Custom tags. people might like to use those to sort aswell manually in a way.
that way people can sort out for example what they are waiting for to be done. when they are waiting for BD to be done. and ofc adult overwrites either ways as tags can be a bit iffy at times there i have seen.

ofc custom tags would have little to no limits when used fully.

**Comment by da3dsoul** [comment]
Good idea

**Comment by AreFinstad** [comment]
admin confirmed links for tvdb/trakt seasons and episodes should also be handled i think. not just seasons and episodes. if its not confirmed then ofc it could be wrong. something we dont want.

**Comment by deusxanime** [comment]
Adding comment here per discussion on discord...

Feature request for renamer utility: When using manual renamer in JMM/Shoko Client, allow selecting (via checkbox or otherwise) individual episodes/files rather than an entire series for renaming. Sometimes you only need/want to rename a few episodes inside a series, especially if you have changed another file's name that doesn't necessarily comply with your normal renaming script.

**Comment by da3dsoul** [comment]
It seems it wasn't actually listed hear. I have too much of this memorized. 
We decided to go with Lua for at least the drop folders, renaming doesn't need something that expansive, but it's easy to do. I wanted a script language so that complex control flow didn't fail miserably. Lua is easy to restrict DSL, easy to implement and use, and we can even make a nice IDE with errors. That'll make things easy easier for everyone. Plus, if someone wanted to use a different program to write the script, then Lua has highlighting in most programs.

**Comment by da3dsoul** [comment]
Js was second choice, but DSL and security in general is a pain when dealing with locally run js.

There will also be a wizard to generate scripts based on some of the issues listed above. (Folders by type, multiple mapped drop source and destinations, etc)

**Comment by ElementalCrisis** [comment]
Sounda like we're gonna need to think of a new name for the utility as its going to go beyond file renaming.

**Comment by da3dsoul** [comment]
Well, I think they should be separate (for simplicity and to avoid total disaster), but they can both use all of the same syntax and whatnot. There will just be a difference in the in and out of it.

**Comment by ElementalCrisis** [comment]
Good point, do you see the drop destination scrips being apart of the managing import folders section?

**Comment by da3dsoul** [comment]
Probably links to open forms for either the wizard or full scale editor

**Comment by da3dsoul** [comment]
But yes

**Comment by da3dsoul** [comment]
The webui might only have a wizard and uploader for the script, while desktop has the full editor.

**Comment by Cazzar** [comment]
Current ideas for the design:

on rename it goes through the following setup:
1. Work out what folder we are dropping the media in. Be this at the anime or episode level.
2. Work out if the drop destination has a file naming script.
3. 
    * If there is a folder-specific script run this to determine the new filename
    * Else run a fallback script which is folder-agnostic

only one script will be ran in step 3.

**Comment by bond4u** [comment]
:+1: I'd like to sort my anime by year.
DEST/{year}/{series}/
For example 70-ies to one hdd, 80-ies to another and so on.

**Comment by da3dsoul** [comment]
Yep year is on the list

**Comment by da3dsoul** [comment]
Just had a cool idea that might be useful for people like @Cazzar, who have a ton of drives and split anime among them.
Series size (num of episodes/files)
Split series files by dubbed or not (in case you want both)

**Comment by bigretromike** [comment]
What is current status of this feature? is it: todo/feature-request, wip or implemented?

**Comment by da3dsoul** [comment]
TODO

**Comment by da3dsoul** [comment]
This isn't something that should be done along side another major feature, due to the fact that there will likely be a lot of bugs with it at first

**Comment by bigretromike** [comment]
like with everything that is added new. That is the way this works. 
Thanks for status info, is been quite some time - and a lot of great ideas;

**Comment by da3dsoul** [comment]
True but this deals with importing, and there are so many edge cases it takes awhile to get sorted out

**Comment by Cazzar** [comment]
Step 1 6f907d52bad6ef9e9f4d890e7df78adafe3ed6f6

**Comment by Cazzar** [comment]
ab2b49df628d8b84ec1b5ebf94bb5bfa972d5097 as well,

Most of the base work is now done, I am adding abilities in Desktop to select the "Renamer Engine" to use.

**Comment by Cazzar** [comment]
![image](https://user-images.githubusercontent.com/994214/28239473-440bc4f2-69b0-11e7-8dfc-1e823a69e98a.png)
I'll just leave this here.

**Comment by bigretromike** [comment]
Any news guys @Cazzar @da3dsoul  ?

**Comment by ElementalCrisis** [comment]
It's on hold at the moment. @cazzar has a workaround with a custom .dll file but I don't see this getting implemented until Desktop 2 as right now it would be pointless to redo the Renamer UI in Desktop.

**Comment by da3dsoul** [comment]
Not really. I'm busy with other stuff. It's basically done in the back end. All that would be needed is the script handler and lots of error checking. The front end will be a lot of work, as I've big plans for it.

**Comment by da3dsoul** [comment]
Okay, so I plan to implement an event system.
My basic plan is as such:
You can have multiple Rename and FileMove handlers, and settings will need to define an order and enabled status for them (new settings will make that much easier).
There will be pre and post events for if you need to do cleanup or rearrange things. You could fire a script outside of shoko using these. I know that has been requested.
The events will be cancelable, and if cancelled, the move or rename will not occur. The events will have a Result. If the result is null, it'll continue to the next registered event.

How do you guys think we should handle event subscription?
One idea is to make an interface for each, and on init it'll register the events by interface.
IDK what else to guarantee registration of a plugin class.

**Comment by bond4u** [comment]
If post-event fails, will You undo the file move? I'd say post-event Result is irrelevant.
I've been thinking about dedicated filesystem queue. So that hashing etc and file move don't happen in parallel.

**Comment by da3dsoul** [comment]
No, if it fails, it'll continue and try the next one. If it's cancelled (an active action from the plugin), then it won't continue.

**Comment by Cazzar** [comment]
Renamer plugins work for this now.
